Memcached Receiver Skeleton

keitwb opened the pull request:
This is the start of a basic memcached receiver.  It scrapes some basic
metrics from a memcached instance and sends them along.  The primary
test of the metric generation logic is in the integration test due to
the relative opacity of the memcached protocol and the lack of any
preexisting mock (the integration test uses a real memcached instance
via Docker containers).

There are only five basic metrics to start with but this can be easily
expanded with this framework.

jrcamp commented:
Can you use the new scraper helper? https://github.com/open-telemetry/opentelemetry-collector/blob/015827f6ed982d15c043319c7c367f9431664178/receiver/receiverhelper/scraper.go (It's not been used widely yet so would be good to give it a test drive and see if it needs any tweaking)

anuraaga commented on receiver/memcachedreceiver/config.go:
```diff
@@ -0,0 +1,33 @@
+	// The target memcached host.
+	Endpoint string `mapstructure:"endpoint"`
+
+	Timeout time.Duration `mapstructure:"timeout"`
```
```suggestion
        // Timeout for requests to fetch memcache stats.
	Timeout time.Duration `mapstructure:"timeout"`
```

anuraaga replied:
Sorry you probably shouldn't actually apply the suggestion due to spaces vs tabs but just copy :)

anuraaga commented on receiver/memcachedreceiver/factory.go:
```diff
@@ -0,0 +1,62 @@
+			NameVal: typeStr,
+		},
+		CollectionInterval: 10 * time.Second,
+		Timeout:            10 * time.Second,
```
Should the default timeout be at least a little less than the interval to prevent concurrent requests?

keitwb replied:
Not necessarily because the for/select loop + time.Ticker smooths things out and prevents concurrent requests.

anuraaga reviewed (approved):
Just some small points, thanks!

keitwb commented:
> Can you use the new scraper helper?

What advantage does it provide, especially when there is only one collection loop?  If the ticker loop is desirable to reuse, we can do something like `utils.RunOnInterval` that was in the Smart Agent.

tigrannajaryan commented:
@keitwb @bogdandrutu is this review still in progress?

keitwb commented:
Yes, switched it to use the scraper interface.

I still think you are going to be  far better on long term by just offering helpers that you can compose into the receivers (e.g. the config options, the timer sharing thing, and whatever else) instead of forcing inheritance through the scraper interface (that thing is basically an abstract class with the ScraperBase interface being the abstract methods that must be implemented).  But I'm on to a different team soon, so I'll just share my opinion on it and let you all decide how to proceed with Scrapers and the porting of the monitors.

keitwb commented on cmd/otelcontribcol/components.go:
```diff
@@ -103,6 +104,7 @@ func components() (component.Factories, error) {
 		splunkhecreceiver.NewFactory(),
 		statsdreceiver.NewFactory(),
 		wavefrontreceiver.NewFactory(),
+		memcachedreceiver.NewFactory(),
```
Removed from components.

tigrannajaryan commented:
Tests failed.

keitwb commented:
Coverage does not take into account the integration test.

bogdandrutu commented on internal/common/testing/container/container.go:
```diff
@@ -45,20 +45,28 @@ type Containers struct {
-func (c *Container) AddrForPort(port uint16) string {
-	for dockerPort, bindings := range c.ports {
+// AddrForPort returns the host and port for the container port.
+func (c *Container) AddrForPort(port uint16) (string, uint) {
```
Please split this change in a separate PR (everything in internal/common)

keitwb replied:
Removed it as it isn't very essential.

bogdandrutu commented on receiver/memcachedreceiver/config.go:
```diff
@@ -0,0 +1,33 @@
+	configmodels.ReceiverSettings            `mapstructure:",squash"`
+	receiverhelper.ScraperControllerSettings `mapstructure:",squash"`
+
+	// The target memcached host.
+	Endpoint string `mapstructure:"endpoint"`
```
Consider to use one of the addresses from `confignet` if possible?

keitwb replied:
Done.

bogdandrutu commented on receiver/memcachedreceiver/config.go:
```diff
@@ -0,0 +1,33 @@
+	// The target memcached host.
+	Endpoint string `mapstructure:"endpoint"`
+
+	// The timeout for the memcache stats request
```
s/The timeout/Timeout

keitwb replied:
Done.

bogdandrutu commented on receiver/memcachedreceiver/config.go:
```diff
@@ -0,0 +1,33 @@
+	configmodels.ReceiverSettings            `mapstructure:",squash"`
+	receiverhelper.ScraperControllerSettings `mapstructure:",squash"`
+
+	// The target memcached host.
```
[Go comments](https://golang.org/doc/effective_go.html#commentary) should start with the name of the exported var/func/struct/etc.

keitwb replied:
Done.

benkeith-splunk committed:
Memcached Receiver Skeleton

This is the start of a basic memcached receiver.  It scrapes some basic
metrics from a memcached instance and sends them along.  The primary
test of the metric generation logic is in the integration test due to
the relative opacity of the memcached protocol and the lack of any
preexisting mock (the integration test uses a real memcached instance
via Docker containers).

There are only five basic metrics to start with but this can be easily
expanded with this framework.
